cni/network/invoker_cns.go

```diff
@@ -130,97 +131,143 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
 			hostSubnet:         response.PodIPInfo[i].HostPrimaryIPInfo.Subnet,
 			hostPrimaryIP:      response.PodIPInfo[i].HostPrimaryIPInfo.PrimaryIP,
 			hostGateway:        response.PodIPInfo[i].HostPrimaryIPInfo.Gateway,
-		}
-
-		// set the NC Primary IP in options
-		// SNATIPKey is not set for ipv6
-		if net.ParseIP(info.ncPrimaryIP).To4() != nil {
-			addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+			addressType:        response.PodIPInfo[i].AddressType,
+			macAddress:         response.PodIPInfo[i].MacAddress,
+			isDefaultInterface: response.PodIPInfo[i].IsDefaultInterface,
+			routes:             response.PodIPInfo[i].Routes,
 		}
 
 		logger.Info("Received info for pod",
-			zap.Any("ipv4info", info),
+			zap.Any("ipinfo", info),
 			zap.Any("podInfo", podInfo))
-		ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
-		if ip == nil {
-			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-		}
 
-		ncgw := net.ParseIP(info.ncGatewayIPAddress)
-		if ncgw == nil {
-			// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+		switch info.addressType {
+		case cns.Secondary:
+			ip, ipnet, err := response.PodIPInfo[i].PodIPConfig.GetIPNet()
```

Comment:
i would suggest to move to different function for each case as function is growing big

Reply:
not addressing this?

Reply:
updated